File: backend/app/schemas/audit.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class AuditLogRead(BaseModel):
    id: UUID
    user_id: UUID | None
    device_id: UUID | None
    operation_id: UUID | None
    action: str
    details: dict | None
    ip_address: str | None
    record_hash: str
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditLogFilter(BaseModel):
    device_id: UUID | None = None
    user_id: UUID | None = None
    action: str | None = None
    from_date: datetime | None = None
    to_date: datetime | None = None


# ── Audit review / policy schemas ──────────────────────────────────────────────

class ReviewRequest(BaseModel):
    approved: bool
    comment: str = ""


class AuditPolicyRead(BaseModel):
    id: UUID
    scope_type: str
    scope_id: str
    intent: str
    requires_approval: bool
    updated_at: datetime

    model_config = {"from_attributes": True}


class AuditPolicyUpsert(BaseModel):
    scope_type: str   # "role" | "user"
    scope_id: str     # role name or user UUID
    intent: str
    requires_approval: bool


class AuditOperationRead(BaseModel):
    """Operation enriched with requester and device info for the audit page."""
    id: UUID
    natural_language_input: str
    intent: str | None
    action_plan: dict | None
    status: str
    error_message: str | None
    review_comment: str | None
    reviewed_at: datetime | None
    executed_direct: bool
    created_at: datetime
    updated_at: datetime
    # Enriched fields
    requester_name: str | None = None
    requester_email: str | None = None
    device_name: str | None = None
    device_vendor: str | None = None
    reviewer_name: str | None = None


class UserForPolicyRead(BaseModel):
    id: UUID
    name: str
    email: str
    role: str

    model_config = {"from_attributes": True}

File: frontend/src/types/audit.ts
export interface AuditLog {
  id: string;
  user_id: string | null;
  device_id: string | null;
  operation_id: string | null;
  action: string;
  details: Record<string, unknown> | null;
  ip_address: string | null;
  record_hash: string;
  created_at: string;
}

export interface AuditOperation {
  id: string;
  natural_language_input: string;
  intent: string | null;
  action_plan: Record<string, unknown> | null;
  status: string;
  error_message: string | null;
  review_comment: string | null;
  reviewed_at: string | null;
  executed_direct: boolean;
  created_at: string;
  updated_at: string;
  requester_name: string | null;
  requester_email: string | null;
  device_name: string | null;
  device_vendor: string | null;
  reviewer_name: string | null;
}

export interface AuditPolicy {
  id: string;
  scope_type: "role" | "user";
  scope_id: string;
  intent: string;
  requires_approval: boolean;
  updated_at: string;
}

export interface AuditPolicyUpsert {
  scope_type: "role" | "user";
  scope_id: string;
  intent: string;
  requires_approval: boolean;
}

export interface ReviewRequest {
  approved: boolean;
  comment: string;
}

export interface UserForPolicy {
  id: string;
  name: string;
  email: string;
  role: string;
}

export const AUDIT_INTENTS: { key: string; label: string; group: string; defaultApproval: boolean }[] = [
  { key: "create_rule",              label: "Criar Regra de Firewall",          group: "Firewall",  defaultApproval: true  },
  { key: "edit_rule",                label: "Editar Regra de Firewall",          group: "Firewall",  defaultApproval: true  },
  { key: "delete_rule",              label: "Excluir Regra de Firewall",         group: "Firewall",  defaultApproval: true  },
  { key: "create_nat_policy",        label: "Criar Política NAT",                group: "NAT",       defaultApproval: true  },
  { key: "delete_nat_policy",        label: "Excluir Política NAT",              group: "NAT",       defaultApproval: true  },
  { key: "create_route_policy",      label: "Criar Rota Estática",               group: "Rotas",     defaultApproval: true  },
  { key: "delete_route_policy",      label: "Excluir Rota Estática",             group: "Rotas",     defaultApproval: true  },
  { key: "create_group",             label: "Criar Grupo de Endereços",          group: "Objetos",   defaultApproval: true  },
  { key: "configure_content_filter", label: "Configurar Content Filter",         group: "Segurança", defaultApproval: true  },
  { key: "configure_app_rules",      label: "Configurar App Rules",              group: "Segurança", defaultApproval: true  },
  { key: "add_security_exclusion",   label: "Adicionar Exclusão de Segurança",   group: "Segurança", defaultApproval: false },
  { key: "toggle_gateway_av",        label: "Toggle Gateway Anti-Virus",         group: "Serviços",  defaultApproval: false },
  { key: "toggle_anti_spyware",      label: "Toggle Anti-Spyware",               group: "Serviços",  defaultApproval: false },
  { key: "toggle_ips",               label: "Toggle IPS",                        group: "Serviços",  defaultApproval: false },
  { key: "toggle_app_control",       label: "Toggle App Control",                group: "Serviços",  defaultApproval: false },
  { key: "toggle_geo_ip",            label: "Toggle Geo-IP Filter",              group: "Serviços",  defaultApproval: false },
  { key: "toggle_botnet",            label: "Toggle Botnet Filter",              group: "Serviços",  defaultApproval: false },
  { key: "toggle_dpi_ssl",           label: "Toggle DPI-SSL",                    group: "Serviços",  defaultApproval: false },
  { key: "direct_ssh",               label: "Comando SSH Direto",                 group: "Técnico",   defaultApproval: true  },
];
